cocotb/runner.py

```diff
@@ -0,0 +1,721 @@
+        for path in self.python_search:
+            self.env["PYTHONPATH"] += os.pathsep + path
+
+        self.env["PYTHONHOME"] = sysconfig.get_config_var("prefix")
+
+        self.env["TOPLEVEL"] = self.hdl_topmodule
+        self.env["MODULE"] = self.module
+
+        if not os.path.exists(self.build_dir):
+            os.makedirs(self.build_dir)
+
+    def build_command(self):
+        raise NotImplementedError()
+
+    def test_command(self):
+        raise NotImplementedError()
+
+    def build(
+        self,
+        library_name="work",
+        verilog_sources=[],
+        vhdl_sources=[],
+        includes=[],
+        defines=[],
+        parameters={},
+        extra_args=[],
+        hdl_topmodule=None,
+        always=False,
+        build_dir="sim_build",
+        work_dir=None,
+    ):
+        self.init_dir(build_dir, work_dir)
+        os.makedirs(self.build_dir, exist_ok=True)
+
+        # note: to avoid mutating argument defaults, we ensure that no value
+        # is written without a copy. This is much more concise and leads to
+        # a better docstring than using `None` as a default in the parameters
+        # list.
+        self.library_name = library_name
+        self.verilog_sources = self.get_abs_paths(verilog_sources)
+        self.vhdl_sources = self.get_abs_paths(vhdl_sources)
+        self.includes = self.get_abs_paths(includes)
+        self.defines = list(defines)
+        self.parameters = dict(parameters)
+        self.compile_args = extra_args
+        self.always = always
+        self.hdl_topmodule = hdl_topmodule
+
+        for e in os.environ:
+            self.env[e] = os.environ[e]
+
+        cmds = self.build_command()
+        self.execute(cmds)
+
+    def test(
+        self,
+        py_module,
+        hdl_topmodule,
+        toplevel_lang="verilog",
+        testcase=None,
+        seed=None,
+        python_search=[],
+        extra_args=[],
+        plus_args=[],
+        extra_env=None,
+        waves=None,
+        gui=False,
+        build_dir="sim_build",
+        work_dir=None,
+    ):
+
+        self.init_dir(build_dir, work_dir)
+
+        if isinstance(py_module, str):
+            self.module = py_module
+        else:
+            self.module = ",".join(py_module)
+
+        # note: to avoid mutating argument defaults, we ensure that no value
+        # is written without a copy. This is much more concise and leads to
+        # a better docstring than using `None` as a default in the parameters
+        # list.
+        self.python_search = list(python_search)
+        self.hdl_topmodule = hdl_topmodule
+        self.toplevel_lang = toplevel_lang
+        self.sim_args = extra_args
+        self.plus_args = list(plus_args)
+        self.env = dict(extra_env) if extra_env is not None else {}
+
+        if testcase is not None:
+            self.env["TESTCASE"] = testcase
+
+        if seed is not None:
+            self.env["RANDOM_SEED"] = str(seed)
```

Comment:
Some are very handy like `WAVES` and `GUI` so one can debug without changing the code.

`COCOTB_RESULTS_FILE` is special. In `cocotb-test` I used it in `pytest` plugin to combine the results see: https://github.com/themperek/cocotb-test/blob/master/cocotb_test/plugin.py . Maybe there is a better way?

The rest can probably go.

Reply:
How about we change WAVES and GUI to COCOTB_WAVES and COCOTB_GUI? Or is there a way to pass those using pytest commandline switches like `pytest --cocotb-waves --cocotb-gui` like what pytest-coverage does?

Reply:
> Or is there a way to pass those using pytest commandline switches like `pytest --cocotb-waves --cocotb-gui` like what pytest-coverage does?

This can be done but will need those variables in any case (as I know).

Reply:
I like `WAVES` and `GUI` more. This is just for runner not for `cocotb` itself. But no strong opinion.